### ISSUE TITLE
lsp: Request textDocument/hover failed on Cyrillic characters

### ISSUE THREAD

**Issue by koddr** [a-h/templ#482]
Hi, @a-h 

I've encountered a bug in the [VS Code extension](https://github.com/templ-go/templ-vscode) (latest, `v0.0.19`).

If you drag the cursor over a string with Cyrillic characters (runes?) in `*.templ` file, an error pops up:

```console
Request textDocument/hover failed.
Message: rune error
Code: 0 
```

Any text in another language does not cause such trouble.

Look this small screencast to understand:

https://github.com/a-h/templ/assets/11155743/3c1aeaf2-add4-4eab-b493-8bfb8ebc841d

What can I do to prevent this error from appearing? 🤔

**Comment by a-h** [comment]
I can reproduce it with Chinese too.

Hovering over the Chinese text in `var happyBirthday = "生日快乐"` causes a "rune error" too.

Since the underlying error is a `*jsonrpc2.Error` and the message isn't in templ, I figured it must be deeper in the stack and tracked it to gopls: https://cs.opensource.google/go/x/tools/+/master:gopls/internal/golang/hover.go;drc=c3f60b7871c6c0f1f49094d63dcf73622a1d1ea0;l=630

From what I can see so far, it looks like our RangeWriter is using UTF8 positions, when the LSP expects "somewhat" UTF16 positions. Since we're using UTF8 positions, we have 3 bytes per rune, which means gopls starts grabbing weird slices of bytes and doesn't like it.

I think the RangeWriter is writing the "end" of its byte sequence position, rather than the start of its position, or something like that.

**Comment by a-h** [comment]
In Neovim, it's possible to enable additional logging with `:lua vim.lsp.set_log_level("DEBUG")`.

With that in place, you can use `:LspLog` to see what's going on.

My test file is:

```go
package main

var nihao = "你好"

var 你好 = "nihao"
```

I'm putting the cursor on `你` (ni) on the last line and using `gr` (go to reference).

Here's the log data: 

```
[DEBUG][2024-04-22 10:48:10] .../lua/vim/lsp.lua:1391	"LSP[gopls]"	"client.request"	2	"textDocument/references"	{  context = {    includeDeclaration = true  },  position = {    character = 5,    line = 4  },  textDocument = {    uri = "file:///home/adrian-hesketh/lsptest/main.go"  }}	<function 1>	1
[DEBUG][2024-04-22 10:48:10] .../vim/lsp/rpc.lua:284	"rpc.send"	{  id = 3,  jsonrpc = "2.0",  method = "textDocument/references",  params = {    context = {      includeDeclaration = true    },    position = {      character = 5,      line = 4    },    textDocument = {      uri = "file:///home/adrian-hesketh/lsptest/main.go"    }  }}
[DEBUG][2024-04-22 10:48:10] .../vim/lsp/rpc.lua:387	"rpc.receive"	{  id = 3,  jsonrpc = "2.0",  result = { {      range = {        ["end"] = {          character = 6,          line = 4        },        start = {          character = 4,          line = 4        }      },      uri = "file:///home/adrian-hesketh/lsptest/main.go"    } }}
```

The interesting bit is the `rpc.send` which shows where Neovim thinks my cursor is:

```
position = {      character = 5,      line = 4    }
```

The next character along is char 6, on line 4. So `gopls` is happily using cursor positions, not byte positions. I suspect then, that the RangeWriter needs to stop counting the size of bytes it's written out, and count the number of runes it has output instead.

**Comment by a-h** [comment]
I think I've resolved this in #712 - care to test out the branch and check that it properly resolves it for you @koddr?
